### ISSUE TITLE
BUG: Trailing space when users copy their temporary password from their email is causing log in errors.

### ISSUE THREAD

**Issue by ttlenard** [ustaxcourt/ef-cms#8698]
**Describe the Bug**
We have received support tickets from public users that state they receive the username/password error upon logging in with a temporary password, and they insist that they copied over their information directly from the DAWSON email that is sent when a user's account is created for them.

Upon investigation, we are seeing that some users "grab" a trailing space with the temporary password in their email, and that trailing space is getting recognized as a space in the password field - causing the log in error message.  

We would like to ensure that if there is a trailing space at the end of the password, it is not recognized.
  
**Business Impact/Reason for Severity**
low

**In which environment did you see this bug?**
Prod, test

**Who were you logged in as?**
Any external user

**What were you doing when you discovered this bug? (Using the application, demoing, smoke tests, testing other functionality, etc.)**
This was reported to us via support ticket

**To Reproduce**
1. Log in as an admissions clerk
2. Click on Advanced, and select practitioner
3. Find a practitioner that does not yet have an email associated with their account
4. Add in an email that you have access to
5. Save the practitioner record
6. Log out of DAWSON
7. Next, go to your email
8. Find the account created email that was sent by DAWSON
9. Click the log in link
10. Add in your email
11. Copy the temporary password from the email (and be sure that you "accidentally" grab that trailing space at the end of the temp PW)
12. Paste it into the PW field
13. Click Log in
14. Receive error

**Expected Behavior**
Trailing space shouldn't be counted as a character in the PW

**Actual Behavior**
A clear and concise description of what actually happened.
The trailing space is being counted as a character in the PW field

**Screenshots**

This is just an example, but notice the extra space at the end of the PW that got "copied" over and created the log in error.
<img width="589" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBNldZQkE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--75be36250a28c695b1dda5c11049ff2672f3dc26/image.png" alt="image.png" />

**Desktop (please complete the following information):**

**Smartphone (please complete the following information):**

**Cause of Bug, If Known**

**Process for Logging a Bug:**
* Complete the above information
* Add a severity tag (Critical, High Severity, Medium Severity or Low Severity). See below for priority definition. 

**Severity Definition:**
* Critical Defect
Blocks entire system's or module’s functionality
No workarounds available
Testing cannot proceed further without bug being fixed.

* High-severity Defect
Affects key functionality of an application
There's a workaround, but not obvious or easy
App behaves in a way that is strongly different from the one stated in the requirements

* Medium-severity Defect
A minor function does not behave in a way stated in the requirements.
Workaround is available and easy

* Low-severity Defect
Mostly related to an application’s UI
Doesn't need a workaround, because it doesn't impact functionality

## Definition of Ready for Bugs(Created 10-4-21)
Definition used: A failure or flaw in the system which produces an incorrect or undesired result that deviates from the expected result or behavior. (Note: Expected results are use cases that have been documented in past user stories as acceptance criteria and test cases, and do not include strange behavior unrelated to use cases.)

The following criteria must be met in order for the development team to begin work on the bug.

The bug must:
 - Be focused on solving a user problem
 - Contain data for all fields in the bug template, so the team can pick it up and begin working immediately

Process: If the unexpected results are new use cases that have been identified, but not yet built, new acceptance criteria and test cases should be captured in a new user story and prioritized by the product owner. 

If the Court is not able to reproduce the bug, add the “Unable to reproduce” tag. This will provide visibility into the type of support that may be needed by the Court. In the event that the Court cannot reproduce the bug, the Court will work with Flexion to communicate what type of troubleshooting help may be needed.

## Definition of Done (Updated 4-14-21)
**Product Owner**
 - [ ]  Bug fix has been validated in the Court's test environment

**Engineering**
 - [ ] Automated test scripts have been written
 - [ ] Field level and page level validation errors (front-end and server-side) integrated and functioning
 - [ ] Verify that language for docket record for internal users and external users is identical
 - [ ] New screens have been added to pa11y scripts
 - [ ] All new functionality verified to work with keyboard and macOS voiceover https://www.apple.com/voiceover/info/guide/_1124.html 
 - [ ] READMEs, other appropriate docs and swagger/APIs fully updated
 - [ ] UI should be touch optimized and responsive for external only (functions on supported mobile devices and optimized for screen sizes as required)
 - [ ] Interactors should validate entities before calling persistence methods
 - [ ] Code refactored for clarity and to remove any known technical debt
 - [ ] Deployed to the Court's `test` environment if prod-like data is required. Otherwise, deployed to any `experimental` environment for review.

**Comment by Mwindo** [comment]
I picked this up as a quick interstitial work item but realized it isn't _quite_ as simple as I thought (e.g., using `.trimEnd()` to remove trailing spaces). What if the user specifies a trailing space in their password? Do we disallow this? If so, how do we communicate this rule to the user?

**Comment by ttlenard** [comment]
@Mwindo - We do communicate this with the user when they create an account, but there isn't any messaging on the log in page.

<img width="823" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBK0dPQmc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--9d0a459680473cc4c4017323152aca871f4bffbc/image.png" alt="image.png" />

**Comment by Mwindo** [comment]
@ttlenard Somehow I totally missed this! (I saw "Must contain special character or space" but my brain ignored the first message.) Thank you :)

**Comment by Mwindo** [comment]
The _best_ solution to this bug is to send a link rather than a temporary password. This is difficult: the one-time password and email logic is handled by Cognito, not us.

Moving on, then: The _easiest_ solution to this bug is to trim leading and trailing spaces from passwords before submitting to our authenticator.

There were concerns about 1) users who have "real" leading or trailing spaces in their passwords before DAWSON enforced no leading or trailing spaces and 2) silently modifying user input.

After doing some research, I have determined that Cognito does not and has not allowed leading/trailing spaces (see https://stackoverflow.com/questions/44522600/space-characters-in-cognito-password and https://stackoverflow.com/questions/58767980/aws-cognito-password-regex-specific-to-aws-cognito). So there should be no users that have "real" leading or trailing spaces in their passwords. (Even if this is mistaken, the worst-case scenario is that some finite handful of users open help tickets, which is better than a continuous stream of help tickets from users accidentally copying-and-pasting a space.)

With regard to the concern around silently modifying user input, I cannot find any consistent advice about when/whether to trim/not trim user input: everyone seems to have different opinions. (Even on our team, I have gotten different preferences.) Given that Gmail trims leading/trailing spaces (I tried logging in both ways successfully), I feel ok trimming the user input and will proceed accordingly.
